File: src/lib/types/kbo.ts
export type GameStatus = "scheduled" | "live" | "final" | "postponed";

export type Position =
  | "투수"
  | "포수"
  | "1루수"
  | "2루수"
  | "3루수"
  | "유격수"
  | "좌익수"
  | "중견수"
  | "우익수"
  | "지명타자";

export interface Team {
  id: string;
  name: string;
  shortName: string;
  city: string;
  stadium: string;
  logoPlaceholder: string; // 팀 색상 hex
  primaryColor: string;
  secondaryColor: string;
}

export interface Standing {
  rank: number;
  team: Team;
  wins: number;
  losses: number;
  draws: number;
  winRate: number;
  gamesBehind: number;
  last10: string; // e.g. "6승4패"
  streak: string; // e.g. "3연승"
  recentForm: ("W" | "L" | "D")[]; // 최근 5경기
}

export interface PlayerSeasonStat {
  playerId: string;
  season: number;
  teamId: string;
  // 타자
  games?: number;
  atBats?: number;
  hits?: number;
  homeRuns?: number;
  rbi?: number;
  avg?: number; // 타율
  obp?: number; // 출루율
  slg?: number; // 장타율
  ops?: number; // OPS
  wOBA?: number; // wOBA
  wRC?: number; // wRC+
  war?: number; // WAR
  babip?: number; // BABIP
  iso?: number; // ISO
  bbPct?: number; // BB%
  kPct?: number; // K%
  // 투수
  era?: number; // ERA
  fip?: number; // FIP
  xFip?: number; // xFIP
  whip?: number; // WHIP
  innings?: number; // IP
  strikeouts?: number; // K
  walks?: number; // BB
  kPer9?: number; // K/9
  bbPer9?: number; // BB/9
  hrPer9?: number; // HR/9
  leftOnBasePct?: number; // LOB%
  groundBallPct?: number; // GB%
  wins2?: number; // 승
  losses2?: number; // 패
  saves?: number; // 세이브
  holds?: number; // 홀드
}

export interface Player {
  id: string;
  name: string;
  teamId: string;
  position: Position;
  number: number;
  age: number;
  birthDate: string;
  throws: "우투" | "좌투" | "양투";
  bats: "우타" | "좌타" | "양타";
  height: number; // cm
  weight: number; // kg
  seasonStat: PlayerSeasonStat;
  careerStats: PlayerSeasonStat[];
}

export interface BoxScoreInning {
  inning: number;
  top: number | null; // 원정팀 득점
  bottom: number | null; // 홈팀 득점
}

export interface BoxScore {
  gameId: string;
  innings: BoxScoreInning[];
  awayRuns: number;
  homeRuns: number;
  awayHits: number;
  homeHits: number;
  awayErrors: number;
  homeErrors: number;
  awayLob: number; // 잔루
  homeLob: number;
}

export interface PlayEvent {
  id: string;
  gameId: string;
  inning: number;
  isTop: boolean; // true = 초
  playIndex: number;
  batter: string;
  pitcher: string;
  description: string;
  eventType:
    | "hit"
    | "homerun"
    | "strikeout"
    | "walk"
    | "out"
    | "error"
    | "stolen_base"
    | "double_play"
    | "run_scored"
    | "pitching_change"
    | "inning_end";
  runsScored: number;
  wpa: number; // Win Probability Added (-1 ~ 1)
  li: number; // Leverage Index
  awayScore: number;
  homeScore: number;
  outs: number;
  runnersOn: string; // e.g. "1,2루" | "만루" | ""
}

export interface WinProbabilityPoint {
  playIndex: number;
  inning: number;
  isTop: boolean;
  awayWinProb: number; // 0 ~ 1
  homeWinProb: number; // 0 ~ 1
  label?: string;
}

export interface Game {
  id: string;
  date: string; // YYYY-MM-DD
  startTime: string; // HH:mm
  awayTeamId: string;
  homeTeamId: string;
  stadium: string;
  status: GameStatus;
  awayScore: number | null;
  homeScore: number | null;
  currentInning?: number;
  isTopInning?: boolean;
  boxScore?: BoxScore;
  playByPlay?: PlayEvent[];
  winProbability?: WinProbabilityPoint[];
  awayStartingPitcher?: string;
  homeStartingPitcher?: string;
  awayWinningPitcher?: string;
  homeWinningPitcher?: string;
  mvp?: string;
}

export interface StatLeader {
  category: string;
  unit: string;
  isLowerBetter?: boolean; // ERA처럼 낮을수록 좋은 지표
  leaders: {
    rank: number;
    playerId: string;
    playerName: string;
    teamId: string;
    teamName: string;
    value: number;
  }[];
}

File: backend/app/schemas/kbo.py
from __future__ import annotations

from enum import Enum
from typing import List, Optional

from pydantic import BaseModel


class GameStatus(str, Enum):
    scheduled = "scheduled"
    pre_game = "pre_game"
    live = "live"
    delayed = "delayed"
    suspended = "suspended"
    final = "final"
    cancelled = "cancelled"
    postponed = "postponed"


class Team(BaseModel):
    id: str
    name_ko: str
    name_en: str
    short_name: str
    home_city: str
    stadium: str
    primary_color: str
    secondary_color: str
    logo_url: Optional[str] = None


class PlayerSeasonStat(BaseModel):
    season: int
    # 타자
    games: Optional[int] = None
    at_bats: Optional[int] = None
    hits: Optional[int] = None
    doubles: Optional[int] = None
    triples: Optional[int] = None
    home_runs: Optional[int] = None
    rbi: Optional[int] = None
    runs: Optional[int] = None
    stolen_bases: Optional[int] = None
    walks: Optional[int] = None
    strikeouts_bat: Optional[int] = None
    avg: Optional[float] = None
    obp: Optional[float] = None
    slg: Optional[float] = None
    ops: Optional[float] = None
    woba: Optional[float] = None
    wrc_plus: Optional[int] = None
    babip: Optional[float] = None
    iso: Optional[float] = None
    war_bat: Optional[float] = None
    # 투수
    wins: Optional[int] = None
    losses: Optional[int] = None
    saves: Optional[int] = None
    holds: Optional[int] = None
    innings_pitched: Optional[float] = None
    era: Optional[float] = None
    whip: Optional[float] = None
    fip: Optional[float] = None
    xfip: Optional[float] = None
    k_per_9: Optional[float] = None
    bb_per_9: Optional[float] = None
    hr_per_9: Optional[float] = None
    lob_pct: Optional[float] = None
    gb_pct: Optional[float] = None
    strikeouts_pit: Optional[int] = None
    war_pit: Optional[float] = None


class Player(BaseModel):
    id: str
    team_id: str
    name_ko: str
    name_en: Optional[str] = None
    position: str
    bats: Optional[str] = None
    throws: Optional[str] = None
    jersey_number: Optional[str] = None
    birth_date: Optional[str] = None
    active_status: bool = True
    season_stat: Optional[PlayerSeasonStat] = None


class BoxScoreLine(BaseModel):
    inning: int
    away_runs: int
    home_runs: int


class BoxScore(BaseModel):
    lines: List[BoxScoreLine]
    away_hits: int
    home_hits: int
    away_errors: int
    home_errors: int
    away_total: int
    home_total: int


class PlayEvent(BaseModel):
    id: str
    game_id: str
    inning: int
    inning_half: str  # "top" | "bottom"
    event_index: int
    batting_team_id: str
    fielding_team_id: str
    batter_id: Optional[str] = None
    pitcher_id: Optional[str] = None
    event_type: str
    description: str
    outs_before: int
    outs_after: int
    base_state_before: str  # e.g. "000", "100", "110"
    base_state_after: str
    away_score_before: int
    home_score_before: int
    away_score_after: int
    home_score_after: int
    rbi: int = 0
    is_scoring_play: bool = False
    is_home_run: bool = False
    wpa: Optional[float] = None
    li: Optional[float] = None


class WinProbabilityPoint(BaseModel):
    play_index: int
    inning: int
    inning_half: str
    home_win_probability: float
    away_win_probability: float
    description: Optional[str] = None


class Game(BaseModel):
    id: str
    season: int
    game_date: str
    start_time: str
    stadium: str
    away_team_id: str
    home_team_id: str
    away_score: int
    home_score: int
    status: GameStatus
    inning: Optional[int] = None
    box_score: Optional[BoxScore] = None
    play_events: Optional[List[PlayEvent]] = None
    win_probability: Optional[List[WinProbabilityPoint]] = None


class Standing(BaseModel):
    rank: int
    team_id: str
    wins: int
    losses: int
    draws: int
    win_rate: float
    games_behind: float
    streak: str
    recent_form: List[str]
    runs_scored: Optional[int] = None
    runs_allowed: Optional[int] = None
    run_differential: Optional[int] = None
    pythagorean_wp: Optional[float] = None


class StatLeaderEntry(BaseModel):
    rank: int
    player_id: str
    player_name: str
    team_id: str
    value: float


class StatLeader(BaseModel):
    category: str
    category_ko: str
    unit: str
    is_lower_better: bool
    leaders: List[StatLeaderEntry]


class SabermetricSnapshot(BaseModel):
    player_id: str
    season: int
    wrc_plus: Optional[int] = None
    war: Optional[float] = None
    babip: Optional[float] = None
    iso: Optional[float] = None
    fip: Optional[float] = None
    xfip: Optional[float] = None
